### ISSUE TITLE
Failure in retrieving a new MODIS product, even after adding it

### ISSUE THREAD

**Issue by Maria-da-Conceicao-Pereira** [MatMatt/MODIS#27]
Hello to all, especially developers,

In my analysis I have to use a new MODIS product, MCD64A1 V006 (Burned Area) but, in spite of the fact that apparently I've added it successfully, the package doesn't retrieve it correctly. I'll explain my procedure and problem below in more detail.

For the purpose of adding this new product I used the function 'addProduct':

`MODIS:::addProduct(product="MCD64A1", sensor="MODIS", platform="Combined", pf1="MOTA", pf2="MCD", topic="Burned Area", type="Tile", res="500m", temp_res="Monthly", server=c("LPDAAC", "LAADS"), path_ext="D:/PhD/R_Project", overwrite=TRUE)`

I managed to create, through this function, the file 'MODIS_Products.RData' (by means of a .cvs file with the structure of the list given originally by 'MODIS:::MODIS_Products'); that file, when loaded in R, corresponds to a list ('MODIS_Products') and the programme reads it correctly, with the new product and all its properties, defined as values of the arguments of the function 'addProduct'. The problem is that, when using the package to read the file (so, with 'MODIS:::MODIS_Products' and not just 'MODIS_Products'), the new product is not included. And in fact in a simple script to download the new product, I always get an error confirming that the new product is not found:

`transDate(begin="2003.01.01",end="2003.04.31")`
`getHdf(product="MCD64A1", begin="2003-01-01", end="2003-04-31", tileH=21, tileV=10)`
`No product found with the name MCD64A1 try 'getProduct()' to list available products`

I even overwrote the original file 'MODIS_Products.RData', that comes with the installation of the package (folder 'external'), but it still doesn't work.

I also tried to unlist the list 'MODIS_Products', so that the function 'getHdf' can read the value 'MCD64A1' directly as character, but this also doesn't solve the problem. And anyway, I feel that this is not the issue because if the new product were added and read correctly by the package, this shouldn't constitute an impediment since the components of the list are in character format, I think.

I also don't think that the issue is about adding a new server or collection, I think that the only function that matters here is 'addProduct'. In relation to the server, as long as the arguments of this function (product, sensor, platform...) are correctly defined, there is no reason why the software shouldn't retrieve the new product; in relation to the collection, I think it's also not necessary to add it because the new product at stake, MCD64A1, only has one collection (006).

I've thus tried everything I consider logical and consequently started to think that there is a bug here. 

If someone reads this message, especially amongst the developers, please help me if you can since I've been busy with this problem already for some time and still haven't been able to solve it, which means that I can't advance in my work. I've also left a comment in issue #20, created by someone with exactly the same problem (but with no solution publicly registered yet).

Thank you in advance!

Greetings,
Maria Pereira

**Comment by fdetsch** [comment]
Hi @Maria-da-Conceicao-Pereira, 

`MODIS:::addProduct` and, similarly, `MODIS:::addServer` and `MODIS:::addCollection` mentioned in issue #20 (thanks for pointing this out) are meant to be used by developers only. As you have correctly noticed, these non-exported helper functions allow us to flexibly add new products (or servers, or collections) to the built-in look-up tables. The next time the package is being built, these new entries will then be available for `getProduct()` or higher-level processing functions like `runGdal()`.

Long story short, I will adopt the above `MODIS:::addProduct` call for [MCD64A1](https://lpdaac.usgs.gov/dataset_discovery/modis/modis_products_table/mcd64a1_v006) and let you know when the changes are available from the [develop](https://github.com/MatMatt/MODIS/tree/develop) branch. In addition, I will probably add those helper functions to `.Rbuildignore` in order to avoid unnecessary confusion in the future.

Best, 
Florian

**Comment by fdetsch** [comment]
Hi again, 

as of commit 9b054e99b6c2fc9ef0d6e0d53c7e0c9b60c7de39, MCD64A1 should be available to all **MODIS** users. @Maria-da-Conceicao-Pereira, could you please run the below code snippet and verify? Please note that until the next official version update on CRAN, these changes will be available from the 'develop' branch only. 

```{r}
devtools::install_github("MatMatt/MODIS", ref = "develop")
library(MODIS)

getProduct("MCD64A1")
# MCD64A1 the Monthly Tile Burned Area product from MODIS Combined with a ground resolution of 500m 

getCollection("MCD64A1", forceCheck = TRUE)
# Updating collections from LPDAAC for platform: Combined 
# $MCD64A1
# [1] "006"

tfs = runGdal("MCD64A1", tileH = 21, tileV = 9
              , begin = "2017001", end = "2017059"
              , SDSstring = "10100", job = "burnedArea")
tfs
# $MCD64A1.006
# $MCD64A1.006$`2017-01-01`
# [1] "/.../MODIS_ARC/PROCESSED/burnedArea/MCD64A1.A2017001.Burn_Date.tif"
# [2] "/.../MODIS_ARC/PROCESSED/burnedArea/MCD64A1.A2017001.QA.tif"       
# 
# $MCD64A1.006$`2017-02-01`
# [1] "/.../MODIS_ARC/PROCESSED/burnedArea/MCD64A1.A2017032.Burn_Date.tif"
# [2] "/.../MODIS_ARC/PROCESSED/burnedArea/MCD64A1.A2017032.QA.tif"       
```

**Comment by Maria-da-Conceicao-Pereira** [comment]
Hello @fdetsch,
first of all thank you for your prompt reply and help.

I installed the package devtools without problems:

`> library(devtools, lib.loc="D:/PhD/R_Project/Library", logical.return=TRUE)
[1] TRUE`

Next I run the code snippet you sent above but I still get an error and the desired product continues to not be recognised:

`> devtools::install_github("MatMatt/MODIS", ref = "develop")
Downloading GitHub repo MatMatt/MODIS@develop
from URL https://api.github.com/repos/MatMatt/MODIS/zipball/develop
Installing MODIS
Installation failed: missing value where TRUE/FALSE needed

> library(MODIS)

> getProduct("MCD64A1")
No product found with the name MCD64A1 try 'getProduct()' to list available products.
NULL

> # MCD64A1 the Monthly Tile Burned Area product from MODIS Combined with a ground resolution of 500m 
> 
> getCollection("MCD64A1", forceCheck = TRUE .... [TRUNCATED] 
No product found with the name MCD64A1 try 'getProduct()' to list available products.
Error in getCollection("MCD64A1", forceCheck = TRUE) : Unknown product`

When I run your script I already had package MODIS installed (with MODIS options defined) and loaded. Is there any problem with that?

I'll wait for your next clarification.

Best greetings,
Maria Pereira

**Comment by fdetsch** [comment]
The installation of the most recent **MODIS** develop version failed in your case, see above error:

> Installation failed: missing value where TRUE/FALSE needed

I ran `devtools::install_github()` on various systems now and it finished every time without any issue. In addition, all the subsequent function calls (ie `getProduct()`, `getCollection()`, `runGdal()`) worked. Maybe you could 

* restart R and try it again; 
* manually uninstall **MODIS** using `remove.packages()` and try it again;
* provide the output of `sessionInfo()` which might provide further insights.

**Comment by Maria-da-Conceicao-Pereira** [comment]
Hello @fdetsch,
sorry for replying only now, in the meanwhile I had to divert to another line of work, which will still keep me busy for a few more days.
I'll implement your suggestions and get back to you, as soon as possible.
Best greetings,
Maria Pereira

**Comment by Maria-da-Conceicao-Pereira** [comment]
Hello @fdetsch,
I've uninstalled everything (R, RStudio and packages), installed everything again, and finally it worked...! I couldn't discover exactly what was the problem, it could have been the fact that the most recent version of devtools was elaborated for a R version slightly more advanced than the one I had installed.
With the develop version of MODIS R installed, the new product is already recognised, an updated data.frame with MODIS collections was created in the .auxiliaries folder (collections.RData) and with this information it is in principle possible to run other functions of the package:

`> getProduct("MCD64A1")
MCD64A1 the Monthly Tile Burned Area product from MODIS Combined with a ground resolution of 500m 
> # MCD64A1 the Monthly Tile Burned Area product from MODIS Combined with a ground resolution of 500m 
> 
> getCollection("MCD64A1", forceCheck = TRUE)
Updating collections from LPDAAC for platform: Combined 
$MCD64A1
[1] "006"

> # Updating collections from LPDAAC for platform: Combined 
> # $MCD64A1
> # [1] "006"`

Indeed, I managed to download already a series of images. Unfortunately, the fact that I'm using now a develop version of the package, doesn't allow me to proceed with the preprocessing of the images, which I need to accomplish so badly! Please see below the outcome in the console when trying to use runGdal:

`> tfs = runGdal("MCD64A1", tileH = 21, tileV = 9
+               , begin = "2017001", end = "2017059"
+               , SDSstring = "10100", job = "burnedArea")
########################
outProj          =  asIn 
pixelSize        =  asIn 
resamplingType   =  near 
Error in rgdal::rawTransform(projfrom, projto, nrow(xy), xy[, 1], xy[,  : 
  no arguments in initialization list
> 
> runGdal(product="MCD64A1", begin="2003001", end="2003335", tileH=21, tileV=10, job="2003", outProj="asIn", pixelSize="asIn",
+         localArcPath="E:/GIS_and_RS_Data_and_Processing/Workspace/MODIS_Paper_1/MODIS_ARC/MODIS/MCD64A1.006",
+         outDirPath="E:/GIS_and_RS_Data_and_Processing/Workspace/MODIS_Paper_1/MODIS_ARC_PROCESSED/MODIS_PP_MCD64A1")
########################
outProj          =  asIn 
pixelSize        =  asIn 
resamplingType   =  near 
Error in rgdal::rawTransform(projfrom, projto, nrow(xy), xy[, 1], xy[,  : 
  no arguments in initialization list`

Going through the manual of the package (rgal) used by this function, I realise that 'rawTransform', together with GDALcall,  are "functions allow more direct access to some of the rgdal C API; these are advanced methods intended for package developers only." And in fact, looking at the arguments of these functions, I think it's out of my reach to apply them. Note that the usage of function rawTransform is rawTransform(projfrom, projto, n, x, y, z=NULL).

I've tried uninstalling and reinstalling the package rgal, the same for MODIS, and one or two other minor options, and nothing works. I wonder how it will be possible to overcome this problem, since for now I need the develop version of MODIS (while it is not updated to accomodate the new product), but at the same time, that very develop version requires functions that, not being a developer, I cannot manage... I'd be very grateful if you could render me your precious help, once more.

Best greetings,
Maria Pereira

**Comment by Maria-da-Conceicao-Pereira** [comment]
Hello @fdetsch,
I forgot to mention above that I also tried to use runMrt, as a way around the error given by runGdal when applied from the develop version of MODIS, but it doesn't seem to me a good option because the version of the tool MRT seems outdated in the LP DAAC website (at least for Windows), and also because in the manual of MODIS R one of the arguments is said to be "not fully implemented", and that's precisely where I get one of the errors when trying to apply the alternative runMrt.
Best greetings,
Maria Pereira

**Comment by IlonaVi** [comment]
Dear @fdetsch,

I get the same error as @Maria-da-Conceicao-Pereira when I run runGdal: 
`d<-runGdal(product="MOD13A1", begin="2014-06-01", end="2014-07-01", tileH = 15, tileV = 2, outProj="asIn")
# Error in rgdal::rawTransform(projfrom, projto, nrow(xy), xy[, 1], xy[,  : no arguments in initialization list`

I removed and re-installed MODIS and rgdal packages and restarted R. It didn't help. 

This is what my MODISoptions() says:
`DOWNLOAD:
_______________
MODISserverOrder : LAADS, LPDAAC 
dlmethod         : auto 
stubbornness     : high 
wait             : 0.5 
quiet            : FALSE 


PROCESSING:
_______________
GDAL           : GDAL 2.2.1, released 2017/06/23 
MRT            : Not available. Use 'MODIS:::checkTools('MRT')' for more information! 
pixelSize      : asIn 
outProj        : asIn 
resamplingType : NN 
dataFormat     : GTiff `

This is what my sessionInfo() shows:
`R version 3.3.3 (2017-03-06)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] MODIS_1.1.0   raster_2.5-8  mapdata_2.2-6 maps_3.2.0    rgdal_1.2-15 
[6] sp_1.2-5     

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13     nloptr_1.0.4     plyr_1.8.4       bitops_1.0-6    
 [5] class_7.3-14     tools_3.3.3      digest_0.6.12    memoise_1.1.0   
 [9] tibble_1.3.4     gtable_0.2.0     lattice_0.20-34  rlang_0.1.2     
[13] shiny_1.0.5      DBI_0.7          parallel_3.3.3   e1071_1.6-8     
[17] withr_2.0.0      devtools_1.13.4  classInt_0.1-24  mapedit_0.3.2   
[21] grid_3.3.3       sf_0.5-5         R6_2.2.2         ptw_1.9-12      
[25] XML_3.98-1.9     udunits2_0.13    ggplot2_2.2.1    magrittr_1.5    
[29] scales_0.5.0     htmltools_0.3.6  units_0.4-6      mime_0.5        
[33] colorspace_1.3-2 xtable_1.8-2     httpuv_1.3.5     RCurl_1.95-4.8  
[37] lazyeval_0.2.0   munsell_0.4.3  
`

And this is what I get from MODIS:::checkTools():
`Checking availabillity of MRT:
  'MRT_HOME' not set/found! MRT is NOT enabled! See: 'https://lpdaac.usgs.gov/tools/modis_reprojection_tool'
Checking availability of 'FWTools/OSGeo4W' (GDAL with HDF4 support for Windows):
   OK, GDAL 2.2.1, released 2017/06/23 found!`

However, this script works and I'm able to create a local subset of data:
`getHdf("MOD13A1", begin="2014061", end="2014071", tileH=15, tileV=2)`

Do you have any advice how to proceed with the problem?

Thank you for your help,

Ilona

**Comment by fdetsch** [comment]
@Maria-da-Conceicao-Pereira, @IlonaVi This should be fixed as of https://github.com/MatMatt/MODIS/commit/bf56291dd60bc4812410480ba2778cb480c5ab14, see also issue https://github.com/MatMatt/MODIS/issues/28. Can you please verify, i.e. `devtools::install_github("MatMatt/MODIS", ref = "develop")` and then re-run your `runGdal()` related code? If this no longer causes problems, I will submit all relevant changes to CRAN immediately... 

@IlonaVi Don't worry about 

> 'MRT_HOME' not set/found!

for now. @MatMatt and me are trying to figure out whether a future support for the [MODIS Reprojection Tool](https://lpdaac.usgs.gov/tools/modis_reprojection_tool) still makes sense. After all, the software has last been updated in [April 2011](https://lpdaac.usgs.gov/sites/default/files/public/mrt4.1_releasenotes_032911.pdf), and moreover, all the things relevant for processing MODIS `.hdf` files are possible through GDAL anyway.

**Comment by Maria-da-Conceicao-Pereira** [comment]
Hello @fdetsch<https://github.com/fdetsch>,

Thank you for your reply. In the meantime I had to look for other solutions, because I really had to move on, and I ended up pre-processing my images with MRT software itself. It was more lengthy and ‘manual’ than R, but it did the job. In the whole process I uninstalled GDAL (I had to install GDAL for python, with other requirements, while exploring a package based on python, pyModis). So, right now I can’t perform the check you ask below, but I will as soon as I can and then get back to you.

Anyway, if you allow me, I’d like to give my view and opinion, as a user, about the appropriateness of supporting MRT in future versions of MODIS R, something you question in your reply, and also mention in issue #31. I didn’t add any comment to the latter, but in fact I run into the same error when I tried to use runMrt. In my view, and based on my recent experience, I think you should make MRT operational in MODIS R (through function runMrt) and continue supporting it. In fact, if someone runs into the problem, as I did, of not being able to use runGdal in MODIS R, and if that problem takes some time to be solved (subject to developers’ availability, naturally), the situation can be quite exasperating, since the alternative options are, as far as I could see, always worse than what the option of using runMrt in MODIS R would be. In fact: i) the software MRT itself has a very helpful and excellent manual and does the job, but I presume that operating it directly, as I did, is more lengthy and ‘manual’ than running it in R - through function runMrt; ii) pyModis, a package to download and process MODIS images based on python, was a complete headache in terms of installation and turned out to be useless for my purpose (which was to mosaic 30 tiles and convert them into .tif), when compared with running MRT directly; iii) the last option to work with MODIS images would be to use GDAL directly in the shell, but although I didn’t explore it, I presume it would always be preferable to work with R, especially if you have to process large numbers of images. I’m not sure if this reasoning stands for other operating systems, but for Windows, which is the one I use, I definitely think it applies. I thought it would be useful for you to know a user’s view about this, especially from someone who couldn’t use runGDAL and was confronted with the different alternatives to that function.

Greetings,
Maria Pereira

Maria Pereira | PhD Candidate
Wageningen University
Laboratory of Geo-information Science and Remote Sensing (www.grs.wur.nl/UK)
maria2.pereira@wur.nl<mailto:maria2.pereira@wur.nl> | T +31(0)317481697 | M +31(0)637405947

**Comment by fdetsch** [comment]
@Maria-da-Conceicao-Pereira, thanks for this comprehensive view, you definitely have a point there. Judging from various **MODIS** package issues related to this topic, it seems like there are still a lot of folks out there relying on MRT. 

@MatMatt, as the creator of `runMrt()`, do you think you can get it going again? :-) There's still plenty of other stuff for me to take care of, and in fact, I've never really used MRT before (which is probably why I tend to underestimate its capabilities every now and then...).
